trin-core/src/ethereum/bridge.rs

```diff
@@ -0,0 +1,227 @@
+
+impl From<serde_json::Error> for BridgeError {
+    fn from(_: serde_json::Error) -> Self {
+        BridgeError::SerializationError
+    }
+}
+
+impl From<ProviderError> for BridgeError {
+    fn from(error: ProviderError) -> Self {
+        match error {
+            ProviderError::JsonRpcClientError(_) => BridgeError::EthClientError,
+            ProviderError::SerdeJson(_) | ProviderError::HexError(_) => {
+                BridgeError::SerializationError
+            }
+            _ => BridgeError::ConnectionFailure,
+        }
+    }
+}
+
+/// A type alias for a `Result` whose `Err` variant holds a `BridgeError`.
+type Result<T> = std::result::Result<T, BridgeError>;
+
+/// An Ethereum bridge client.
+#[async_trait]
+pub trait EthBridge {
+    type Client: PubsubClient;
+
+    /// Requests the balance of an account at a particular block number.
+    async fn get_balance(&self, address: Address, block_number: U64) -> Result<U256>;
+    /// Requests a block (with full transaction objects) by hash.
+    async fn get_block(&self, hash: H256) -> Result<Option<Block<Transaction>>>;
+    /// Requests the code of an account at a particular block number.
+    async fn get_code(&self, address: Address, block_number: U64) -> Result<Bytes>;
+    /// Requests the value of a storage slot at a particular block number.
+    async fn get_storage(
+        &self,
+        address: Address,
+        location: H256,
+        block_number: U64,
+    ) -> Result<H256>;
+    // TODO: Replace parameters with block hash and transaction index.
+    /// Requests the transaction for a given transaction hash.
+    async fn get_transaction(&self, hash: H256) -> Result<Option<Transaction>>;
+    /// Requests the number of transactions sent (nonce) from an account at a particular block number.
+    async fn get_transaction_count(&self, address: Address, block_number: U64) -> Result<U256>;
+    /// Requests the transaction receipt for a given transaction hash.
+    async fn get_transaction_receipt(&self, hash: H256) -> Result<Option<TransactionReceipt>>;
+    /// Subscribes to a stream of new block headers as new blocks are appended to the chain.
+    /// In the case of chain reorganizations, all new block headers will be sent.
+    async fn subscribe_new_block_headers(
+        &self,
+    ) -> Result<SubscriptionStream<'_, Self::Client, Block<TxHash>>>;
+}
+
+/// A Go-Ethereum (Geth) bridge client. Geth clients support a superset of the standard Ethereum
+/// JSON-RPC APIs.
+#[async_trait]
+pub trait GethBridge: EthBridge {
+    type Client: PubsubClient;
+
+    /// Requests the EIP-1186 proof for an account.
+    /// EIP: https://github.com/ethereum/EIPs/issues/1186.
+    async fn get_proof(
+        &self,
+        address: Address,
+        storage_locations: &[H256],
+        block_number: U64,
+    ) -> Result<EIP1186ProofResponse>;
+}
+
+/// An ethers-rs Ethereum bridge client.
+pub struct EthersBridge<T: JsonRpcClient> {
+    inner: Provider<T>,
```

Comment:
`inner`? Something like `provider` would make more sense to me

Reply:
Yeah, this convention came from `ethers-rs`, but I agree that it's not very meaningful. I can rename this to `provider` or `client` along with a doc comment. Not sure if there is a preference for one over the other.